### PR TITLE
feat: add API contact properties update + documents + photos + avatar

### LINKED ISSUE #2716: Fix missing API contact fields store/update
Since #2251 it's not possible to create or update these information with an API call:
| job | string | The job title of the contact. Max 255 characters. |
| company | string | The company which employs the contact. Max 255 characters. |
| food_preferencies | string | The food preferencies of the contact. Max 100000 characters. |
| first_met_information | string | The information (ie where and how) the user has met the contact. Max 1000000 characters. |
| first_met_date | string | The date you first met the contact. Format: 'YYYY-MM-DD'. |
| first_met_date_is_age_based | boolean | Indicates whether the first_met_date is age based or not. |
| first_met_date_is_year_unknown | boolean | Indicates whether we know the year or not. |
| first_met_date_age | integer | The number of years between the first_met_date and the current year. |
| first_met_through_contact_id | integer | The contact whose made the introduction to this person. |
| avatar_url | string | The URL of an external image that would serve as the avatar of the contact. Max 400 characters. |

### LINKED ISSUE #2905: API for avatars
I'd like to sync the Monica contacts through the JSON API, but I don't see a way to upload a new avatar to Monica.

### PR DESCRIPTION
**Pull request by asbiin**
Close #2716
Close #2905

Adds some api routes:
- PUT `/contacts/{contact}/work` to update work informations
- PUT `/contacts/{contact}/food` to update food preferences
- PUT `/contacts/{contact}/introduction` to update the firstMet information
- PUT `/contacts/{contact}/avatar` to update the avatar of a contact
- POST `/documents` to store a new document
- DELETE `/documents/{document}` to destroy a document
- GET `/photos` to get every photos
- GET `/photos/{photo}` to get one photo
- POST `/photos` to upload a photo (`contact_id` required to link it to the photo)
- DELETE `/photos/{photo}` to destroy a photo

### CODE REVIEW

**Comment by asbiin** [comment]
@djaiss I've updated the content of this PR recently, would you like to review it please?

File: app/Helpers/DBHelper.php [hunk 1]
```diff
@@ -39,7 +39,7 @@ public static function testVersion($version)
      */
     public static function getTables()
     {
-        return DB::select('SELECT table_name
+        return DB::select('SELECT table_name as `table_name`
```
Comment (asbiin) [inline app/Helpers/DBHelper.php]: This one is needed on my computer with MariaDB

File: app/Http/Controllers/Contacts/PhotosController.php [hunk 2]
```diff
@@ -42,11 +42,10 @@ public function store(Request $request, Contact $contact): PhotoResource
     {
         $photo = app(UploadPhoto::class)->execute([
             'account_id' => auth()->user()->account->id,
+            'contact_id' => $contact->id,
             'photo' => $request->photo,
         ]);
 
-        $contact->photos()->syncWithoutDetaching([$photo->id]);
```
Comment (asbiin) [inline app/Http/Controllers/Contacts/PhotosController.php]: This line is now in the UploadPhoto class

File: CHANGELOG.md [hunk 3]
```diff
@@ -6,6 +6,8 @@
 
 ### Enhancements:
 
+* Add API to destroy and store documents
```
Comment (djaiss) [inline CHANGELOG.md]:
```suggestion
* Add API methods to destroy and store documents
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -6,6 +6,8 @@
...
+* Add API to manage photos and avatars
```
Comment (djaiss) [inline CHANGELOG.md]:
```suggestion
* Add API methods to manage photos and avatars
```

File: app/Services/Contact/Contact/UpdateContactWork.php [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        return [
+            'account_id' => 'required|integer|exists:accounts,id',
+            'contact_id' => 'required|integer|exists:contacts,id',
+            'job' => 'nullable|string|max:255',
+            'company' => 'nullable|string|max:255',
+        ];
+    }
+
+    /**
+     * Update a contact.
```
Comment (djaiss) [inline app/Services/Contact/Contact/UpdateContactWork.php]:
```suggestion
     * Update the work information of a contact.
```

File: app/Services/Contact/Contact/UpdateContactFoodPreferences.php [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+     * Get the validation rules that apply to the service.
+     *
+     * @return array
+     */
+    public function rules()
+    {
+        return [
+            'account_id' => 'required|integer|exists:accounts,id',
+            'contact_id' => 'required|integer|exists:contacts,id',
+            'food_preferences' => 'nullable|string',
```
Comment (djaiss) [inline app/Services/Contact/Contact/UpdateContactFoodPreferences.php]: Would it be necessary to limit the length of the string, like 255 chars max for instance?

File: app/Services/Contact/Contact/UpdateContactFoodPreferences.php [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+        ];
+    }
+
+    /**
+     * Update a contact.
```
Comment (djaiss) [inline app/Services/Contact/Contact/UpdateContactFoodPreferences.php]:
```suggestion
     * Update the food preferences of the given contact.
```

File: app/Services/Contact/Contact/UpdateContactIntroductions.php [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+                Rule::requiredIf(function () {
+                    return Arr::get($this->data, 'is_date_known', false) && Arr::get($this->data, 'is_age_based', false);
+                }),
+            ],
+            'add_reminder' => 'nullable|boolean',
+        ];
+    }
+
+    /**
+     * Update a contact.
```
Comment (djaiss) [inline app/Services/Contact/Contact/UpdateContactIntroductions.php]:
```suggestion
     * Update the information about how a contact was introduced.
```

File: tests/Api/Contact/ApiContactControllerTest.php [hunk 9]
```diff
@@ -1393,4 +1393,276 @@ public function test_it_gets_me_contact()
...
+
+    public function test_it_sets_food_preferences()
+    {
+        $user = $this->signin();
+        $contact = factory(Contact::class)->create([
+            'account_id' => $user->account_id,
+        ]);
+
+        $response = $this->json('PUT', '/api/contacts/'.$contact->id.'/food', [
+            'food_preferences' => 'Pas de laitages, le lait c\'est le mal',
```
Comment (djaiss) [inline tests/Api/Contact/ApiContactControllerTest.php]: 🤣

File: app/Services/Contact/Avatar/UpdateAvatar.php [hunk 10]
```diff
@@ -54,12 +54,15 @@ public function execute(array $data) : Contact
...
-        // in case of a photo, set the photo as the avatar
-        if ($data['source'] === 'photo') {
-            $contact->avatar_photo_id = $data['photo_id'];
-            $contact->photos()->syncWithoutDetaching([$data['photo_id']]);
+        switch ($contact->avatar_source) {
+            case 'photo':
+            // in case of a photo, set the photo as the avatar
+                $contact->avatar_photo_id = $data['photo_id'];
+                $contact->photos()->syncWithoutDetaching([$data['photo_id']]);
+            break;
```
Comment (djaiss) [inline app/Services/Contact/Avatar/UpdateAvatar.php]: I think we need a better indentation for the two `break` words.

File: app/Services/Contact/Contact/UpdateContactFoodPreferences.php (hunk 6 above)
Comment (asbiin) [reply app/Services/Contact/Contact/UpdateContactFoodPreferences.php]: Ok the column is a longtext, which is - I guess - 65535 max.

File: app/Services/Contact/Avatar/UpdateAvatar.php (hunk 10 above)
Comment (asbiin) [reply app/Services/Contact/Avatar/UpdateAvatar.php]: ok, fixed

**Pull request merged**